From b7ea9bc30a4c71fdfdb8684ea2f3436cd3aa11c0 Mon Sep 17 00:00:00 2001
From: Alex Kiernan <alex.kiernan@gmail.com>
Date: Sat, 30 Dec 2023 15:13:27 +0000
Subject: [PATCH 1/2] Handle non-existent/empty <CARGO_HOME>/registry/src

If remap-debuginfo is set but cargo isn't vendored into
.cargo/registry/src, don't panic:

| thread 'main' panicked at src/core/builder.rs:1795:26:
| std::fs::read_dir(registry_src) failed with No such file or directory (os error 2)

Signed-off-by: Alex Kiernan <alex.kiernan@gmail.com>
Upstream-Status: Submitted [https://github.com/rust-lang/rust/pull/119445]
---
 src/bootstrap/src/core/builder.rs | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/src/bootstrap/src/core/builder.rs b/src/bootstrap/src/core/builder.rs
index cd276674dee6..ae4d70df339f 100644
--- a/src/bootstrap/src/core/builder.rs
+++ b/src/bootstrap/src/core/builder.rs
@@ -1791,15 +1791,19 @@ pub fn cargo(
         if self.config.rust_remap_debuginfo {
             // FIXME: handle vendored sources
             let registry_src = t!(home::cargo_home()).join("registry").join("src");
-            let mut env_var = OsString::new();
-            for entry in t!(std::fs::read_dir(registry_src)) {
+            if registry_src.is_dir() {
+                let mut env_var = OsString::new();
+                for entry in t!(std::fs::read_dir(registry_src)) {
+                    if !env_var.is_empty() {
+                        env_var.push("\t");
+                    }
+                    env_var.push(t!(entry).path());
+                    env_var.push("=/rust/deps");
+                }
                 if !env_var.is_empty() {
-                    env_var.push("\t");
+                    cargo.env("RUSTC_CARGO_REGISTRY_SRC_TO_REMAP", env_var);
                 }
-                env_var.push(t!(entry).path());
-                env_var.push("=/rust/deps");
             }
-            cargo.env("RUSTC_CARGO_REGISTRY_SRC_TO_REMAP", env_var);
         }
 
         // Enable usage of unstable features
-- 
2.39.0

